jupyterhub/roles.py

```diff
@@ -133,29 +101,17 @@ def ignore(scopename):
 def _expand_scope(scopename):
     """Returns a set of all subscopes"""
 
-    scopes = _get_scope_hierarchy()
-    subscopes = [scopename]
-
-    def _expand_subscopes(index):
-
-        more_subscopes = list(
-            filter(lambda scope: scope in scopes.keys(), subscopes[index:])
-        )
-        for scope in more_subscopes:
-            subscopes.extend(scopes[scope])
+    expanded_scope = []
 
-    if scopename in scopes.keys() and scopes[scopename] is not None:
-        subscopes.extend(scopes[scopename])
-        # record the index from where it should check for "subscopes of sub-subscopes"
-        index_for_sssc = len(subscopes)
-        # check for "subscopes of subscopes"
-        _expand_subscopes(index=1)
-        # check for "subscopes of sub-subscopes"
-        _expand_subscopes(index=index_for_sssc)
+    def _add_subscopes(scopename):
+        expanded_scope.append(scopename)
+        if scopes.scope_definitions[scopename].get('subscopes'):
+            for subscope in scopes.scope_definitions[scopename].get('subscopes'):
+                _add_subscopes(subscope)
 
```

Comment:
Smart to reuse the scope definitions this way! Why do you use a nested method here?

Reply:
It's a recursive function that keeps calling itself until the subscopes have no subscopes of their own based on the dictionary (each scope either has a list of its immediate subscopes or it doesn't). It basically walks through every scope in a list of subscopes for the given scope.